Sources/GRPCHTTP2Core/GRPCMessageFramer.swift

```diff
@@ -0,0 +1,104 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import NIOCore
+
+/// A ``GRPCMessageFramer`` helps with the framing of gRPC data frames:
+/// - It prepends data with the required metadata (compression flag and message length).
+/// - It compresses messages using the specified compression algorithm (if configured).
+/// - It coalesces multiple messages (appended into the `Framer` by calling ``append(_:compress:)``)
+/// into a single `ByteBuffer`.
+struct GRPCMessageFramer {
+  /// Length of the gRPC message header (1 compression byte, 4 bytes for the length).
+  static let metadataLength = 5
+
+  /// Maximum size the `writeBuffer` can be when concatenating multiple frames.
+  /// This limit will not be considered if only a single message/frame is written into the buffer, meaning
+  /// frames with messages over 64KB can still be written.
+  /// - Note: This is expressed as the power of 2 closer to 6KB (i.e., 6KiB) because `ByteBuffer`
+  /// reserves capacity in powers of 2. This way, we can take advantage of the whole buffer.
+  static let maximumWriteBufferLength = 16_384
+
+  private var pendingMessages: OneOrManyQueue<PendingMessage>
+
+  private struct PendingMessage {
+    let bytes: [UInt8]
+    let compress: Bool
+  }
+
+  private var writeBuffer: ByteBuffer
+
+  init() {
+    self.pendingMessages = OneOrManyQueue()
+    self.writeBuffer = ByteBuffer()
+  }
+
+  /// Queue the given bytes to be framed and potentially coalesced alongside other messages in a `ByteBuffer`.
+  /// The resulting data will be returned when calling ``GRPCMessageFramer/next()``.
+  /// If `compress` is true, then the given bytes will be compressed using the configured compression algorithm.
+  mutating func append(_ bytes: [UInt8], compress: Bool) {
+    self.pendingMessages.append(PendingMessage(bytes: bytes, compress: compress))
+  }
+
+  /// If there are pending messages to be framed, a `ByteBuffer` will be returned with the framed data.
+  /// Data may also be compressed (if configured) and multiple frames may be coalesced into the same `ByteBuffer`.
+  /// - Throws: If an error is encountered, such as a compression failure, an error will be thrown.
+  mutating func next() throws -> ByteBuffer? {
+    if self.pendingMessages.isEmpty {
+      // Nothing pending: exit early.
+      return nil
+    }
+
+    self.writeBuffer.clear()
+    defer {
+      if self.writeBuffer.capacity > Self.maximumWriteBufferLength {
+        self.writeBuffer = ByteBuffer()
+      }
+    }
+
+    var messagesToCoalesce = 0
+    var requiredCapacity = 0
+    for message in self.pendingMessages {
+      messagesToCoalesce += 1
+      requiredCapacity += message.bytes.count + Self.metadataLength
+    }
+    self.writeBuffer.clear(minimumCapacity: requiredCapacity)
+
+    while messagesToCoalesce > 0 {
```

Comment:
This can be simplified now as well: `while let message = self.pendingMessages.pop()`